src/service/builder.js

```diff
@@ -0,0 +1,236 @@
+ * limitations under the License.
+ */
+
+import {LayoutPriority} from '../layout';
+import {READY_SCAN_SIGNAL} from './resources-interface';
+import {VisibilityState} from '../visibility-state';
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {hasNextNodeInDocumentOrder, isIframed} from '../dom';
+import {removeItem} from '../utils/array';
+
+const ID = 'builder';
+
+/** @implements {../service.Disposable} */
+export class Builder {
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc  */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.observer_ = new win.IntersectionObserver((e) => this.observed_(e), {
+      // Root bounds are not important, so we can use the `root:null` for a
+      // top-level window.
+      root: isIframed(win) ? win.document : null,
+      rootMargin: '250% 31.25%',
+      threshold: 0.001,
+    });
+
+    /** @private @const {!Map<!AmpElement, {asap: boolean, isIntersecting: boolean}>} */
+    this.targets_ = new Map();
+
+    /** @private {?Array<!AmpElement>} */
+    this.parsingTargets_ = [];
+
+    /** @private {boolean} */
+    ampdoc.whenReady().then(() => this.checkParsing_());
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() =>
+      this.docVisibilityChanged_()
+    );
+  }
+
+  /** @override */
+  dispose() {
+    this.observer_.disconnect();
+    this.targets_.clear();
+    if (this.visibilityUnlisten_) {
+      this.visibilityUnlisten_();
+      this.visibilityUnlisten_ = null;
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  scheduleAsap(target) {
+    this.targets_.set(target, {asap: true, isIntersecting: false});
+    this.waitParsing_(target);
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  schedule(target) {
+    if (this.targets_.has(target)) {
+      return;
+    }
+
+    if (target.deferredBuild()) {
+      this.targets_.set(target, {asap: false, isIntersecting: false});
+      this.observer_.observe(target);
+    } else {
+      this.targets_.set(target, {asap: false, isIntersecting: true});
+    }
+
+    this.waitParsing_(target);
+    this.signalScanReady_();
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  unschedule(target) {
+    if (!this.targets_.has(target)) {
+      return;
+    }
+
+    this.targets_.delete(target);
+
+    this.observer_.unobserve(target);
+
+    if (this.parsingTargets_) {
+      removeItem(this.parsingTargets_, target);
+      this.checkParsing_();
+    }
+  }
+
+  /** @private*/
+  signalScanReady_() {
+    if (!this.scheduledReady_) {
+      this.scheduledReady_ = true;
+      const {win} = this.ampdoc_;
+      win.setTimeout(() => {
+        // This signal mainly signifies that some of the elements have been
+        // discovered and scheduled.
+        this.ampdoc_.signals().signal(READY_SCAN_SIGNAL);
```

Comment:
Can you explain why we need to handle this in builder?

Reply:
Currently the `Resources` manager issues this signal after the first element is found. After this PR that code will change to issuing this signal after the first V0 element is found since V1 elements will not go into this branch anymore. To make sure that this signal is sent in all cases, I've duplicated it here. That way a page that only contains V1 elements would still be correctly handled.

Reply:
> Currently the Resources manager issues this signal after the first element is found.

Maybe I'm confused, but thats not what I'm seeing. It looks like it always waits to signal until the full document parse completes & all v0 services are installed (ampInitComplete), and first round of measurements completed.

https://github.com/ampproject/amphtml/blob/fac0f5f629752c204388b7f5fadd1d4aad01ac47/src/service/resources-impl.js#L815-L834

---

Nonblocking for me, but I tend to agree with Justin that this ideally would live elsewhere

Reply:
Yeah. It's a bit of a headscratcher. I'll think about which of these rules are actually needed here since we don't do direct measurements and `InOb` already gives us some protection from sampling too early. Document-fully-parsed part is the one I'm not 100% sure yet about. But I think `Builder` is generally a relevant place to have it: the main reason for it is to signal when the document has started actively scheduling elements.

Reply:
To update on this. I indeed blocked this signal until `DOMReady` to be more like the `Resources` version.